src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -39,22 +39,34 @@ internal static SwitchBinder Create(Binder next, SwitchStatementSyntax switchSyn
                 // for all operations; we use it to enhance test coverage.
                 (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false ||
                  parseOptions?.Features.ContainsKey("testV7SwitchBinder") != false ||
-                 switchSyntax.HasErrors && HasPatternSwitchSyntax(switchSyntax))
+                 switchSyntax.HasErrors && HasPatternSwitchSyntax(next.Compilation, switchSyntax))
                 ? new PatternSwitchBinder(next, switchSyntax)
                 : new SwitchBinder(next, switchSyntax);
         }
 
-        internal static bool HasPatternSwitchSyntax(SwitchStatementSyntax switchSyntax)
+        internal static bool HasPatternSwitchSyntax(CSharpCompilation compilation, SwitchStatementSyntax switchSyntax)
         {
+            var caseExpressions = ArrayBuilder<CSharpSyntaxNode>.GetInstance();
             foreach (var section in switchSyntax.Sections)
             {
-                if (section.Labels.Any(SyntaxKind.CasePatternSwitchLabel))
+                foreach (var label in section.Labels)
                 {
-                    return true;
+                    switch (label.Kind())
+                    {
+                        case SyntaxKind.CasePatternSwitchLabel:
+                            caseExpressions.Free();
+                            return true;
+                        case SyntaxKind.CaseSwitchLabel:
+                            caseExpressions.Add(((CaseSwitchLabelSyntax)label).Value);
+                            continue;
+                    }
                 }
             }
 
-            return false;
+            var dummyBinder = new BuckStopsHereBinder(compilation);
+            var result = ExpressionVariableFinder.CountExpressionVariables(dummyBinder, caseExpressions) != 0;
```

Comment:
>var result = ExpressionVariableFinder.CountExpressionVariables(dummyBinder, caseExpressions) != 0; [](start = 12, length = 98)

This feels too complicated. For example, we could simply add locals to BoundSwitchSection and then propagate them to the ISwitchCaseOperation node in the factory. See an assert I added in SwitchBinder.BindSwitchSection in [https://github.com/dotnet/roslyn/pull/26903](https://github.com/dotnet/roslyn/pull/26903). That should give you an idea how to get to the locals.